### DISCUSSION TITLE
Build Fira Code with stylistic variants

### DISCUSSION THREAD

**Discussion by dsejas** [tonsky/FiraCode#1350] [Q&A] [answered]
Hello! I am trying to use Fira Code in [Doom Emacs](https://github.com/hlissner/doom-emacs), but I would really, really like to use the variant `cv31`. Unfortunately, I couldn't find any way to do that in a Doom-y way. So, I am wondering how could I build Fira Code with the normal parenthesis replaced with that variant.

Thank you very much in advance for your answers!

**Comment by tonsky** [comment, marked as answer]
Try this advice https://github.com/tonsky/FiraCode/issues/1348

**Reply by dsejas** [reply]
Hello! I tried this and got the following error message:

    fontmake: Error: In 'FiraCode.glyphs' -> 'master_ufo/FiraCode.designspace' -> 'master_ufo/instance_ufo/FiraCode-Light.ufo': Compiling UFO failed: master_ufo/instance_ufo/FiraCode-Light.ufo/features.fea:2031:1: Expected glyph class definition or statement: got NAME featureNames
    make: *** [Makefile:7: build] Error 1

Curiously, I get this error using Docker and also following [this instructions](https://github.com/tonsky/FiraCode#building-fira-code-locally).

**Reply by tonsky** [reply]
Interesting! It’s the same error @Mango0x45 was having in #1348. Not sure why it happens, I thought Docker builds were meant to be reproducible

**Reply by tonsky** [reply]
What OS are you trying this on by the way? And what type of CPU?

**Reply by tonsky** [reply]
What exact commands have you run? Which Docker image have you used? Locally built one or `tonsky/firacode:latest`?

**Reply by dsejas** [reply]
Hey, @tonsky! Thank you for your interest in my case. I was able to use Docker with the following steps:

    sudo systemctl stop docker
    sudo rm -rf /var/lib/docker
    sudo rm /var/run/docker.pid
    git clone https://github.com/tonsky/FiraCode.git
    cd FiraCode
    sudo dockerd &
    sudo docker run --rm -v <PATH/TO/FIRACODE/SOURCE/FOLDER>:/opt tonsky/firacode:latest ./script/build.sh

Here, `<PATH/TO/FIRACODE/SOURCE/FOLDER>` must be an absolute path, not relative (for example, `\home\tonsky\fonts\FiraCode`).

The funny part is that yesterday this caused a lot of error messages, but now it doesn't, it just works perfectly. I am letting you know in case you want to document this procedure. Notice this is very similar to @Mango0x45's solution in [#1348](https://github.com/tonsky/FiraCode/issues/1348).

Just in case, here is my system information:

- OS: Kubuntu 21.10
- Docker version: 20.10.7
- Kernel version: 5.13.0-25-generic (64-bit)
- Processor: 4 × Intel® Core™ i5-6200U CPU @ 2.30GHz

However, my problem now is that I cannot figure out how to modify the glyphs file so that it uses the `cv31` variant. Every time I add the code of the "name = cv31" part to the "name = calt" part, I get this error and many others similar:

    ERRO[2022-01-14T14:18:44.215924087-04:00] failed to process event
    container=9ce338cbfb8f31180c6b9f84338750e6120fc8698c95572ac2f4cceffaa09740 error="could not find container
    9ce338cbfb8f31180c6b9f84338750e6120fc8698c95572ac2f4cceffaa09740: No such container:
    9ce338cbfb8f31180c6b9f84338750e6120fc8698c95572ac2f4cceffaa09740" event=create event-info="
    9ce338cbfb8f31180c6b9f84338750e6120fc8698c95572ac2f4cceffaa09740
    9ce338cbfb8f31180c6b9f84338750e6120fc8698c95572ac2f4cceffaa09740 35244 0 0001-01-01 00:00:00 +0000 UTC false <nil>}"
    module=libcontainerd namespace=moby

Could you help me with this part, please?

**Reply by tonsky** [reply]
This looks like Docker error, not build log?

**Reply by dsejas** [reply]
Hello, @tonsky! Indeed, this must have been a problem with my Docker installation, because, after reinstalling it, the error messages are no longer reproducible. However, I am still unable to obtain the cv31 character variant with the procedure you very kindly indicated.

Just in case, I am attaching [my log file](https://github.com/tonsky/FiraCode/files/7893642/FiraCode_Docker.log) from the latest attempt. (I should point out that this is a ~1000 line file.)

On the other hand, out of pure obstinacy, I was able to find an alternative procedure to obtain the cv31 variant of Fira Code, albeit a quite contrive procedure. Here it is in case you would like to document it, comment on it and/or adapt it. (I started with a new session, so no docker instance was running.)

    sudo rm -rf /var/lib/docker
    sudo rm /var/run/docker.pid
    git clone https://github.com/tonsky/FiraCode.git
    cd FiraCode
    sudo dockerd &
    sudo docker run --rm -v <FIRACODE/SOURCE/FOLDER>:/opt tonsky/firacode:latest ./script/build.sh

This should create a folder called "`master_ufo`". Inside, you will find the subfolder `FiraCode-Bold.ufo/glyphs`. Follow the following steps:

1. Open the file `parenleft.gliph` and cut the line that reads "`<unicode hex="0028"/>`". (This is the unicode position of the left parenthesis.) Replace every occurence of the word "`parenleft`" with "`parenleft.cv31`".
2. Paste the line from the previous step into the file `parenleft.cv31.glif` in the same position. Replace every occurrence of "`parenleft.cv31`" with "`parenleft`".
3. Rename `parenleft.glif` as `parenleft.cv31.glif`, and _vice versa_.
4. Repeat steps 1 through 3 for the files `parenright.gliph` and `parenright.cv31.gliph`, and the words "`parenright`" and "`parenright.cv31`".
5. Repeat steps 1 through 4 for the subfolder `FiraCode-Light.ufo/glyphs` inside `master_ufo`.

Now open the file `<FIRACODE/SOURCE/FOLDER>/script/build_variable.sh`, delete the line that reads "`fontmake -g FiraCode_VF.glyphs -o variable --output-dir distr/variable_ttf`" (line 11) and replace it with "`fontmake -m master_ufo/FiraCode.designspace -o variable --production-names --output-dir distr/variable_ttf`". Finally, run the instruction

    sudo docker run --rm -v <PATH/TO/FIRACODE/SOURCE/FOLDER>:/opt tonsky/firacode:latest ./script/build_variable.sh

The file `<FIRACODE/SOURCE/FOLDER>/distr/variable_ttf/FiraCode_VF.ttf` contains the variable font with normal parentheses replaced with the `cv31` variants.

I hope this helps!

**Reply by tonsky** [reply]
Wow, glad you made it work. I would probably edit glyphs file directly, since it’s just one file. It’s not XML or JSON, but pretty simple text format still.

**Reply by dsejas** [reply]
Hello, @tonsky! Sorry for not replying before; my internet connection was intermittent for a few days. Of course you are right! How is it possible that I didn't notice this! Indeed, making the previously described changes directly to `FiraCode.glyphs` makes the trick. Thank you very much, I really appreciate your help!

I am tagging this discussion as "answered".
